include/deal.II/base/symbolic_function.h

```diff
@@ -0,0 +1,315 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2019 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_symbolic_function_h
+#define dealii_symbolic_function_h
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/exceptions.h>
+#include <deal.II/base/function.h>
+#include <deal.II/base/point.h>
+#include <deal.II/base/subscriptor.h>
+#include <deal.II/base/symmetric_tensor.h>
+#include <deal.II/base/tensor.h>
+
+#include <deal.II/differentiation/sd.h>
+
+#include <functional>
+#include <iostream>
+#include <vector>
+
+DEAL_II_NAMESPACE_OPEN
+
+// Forward declarations
+#ifndef DOXYGEN
+template <typename number>
+class Vector;
+namespace Functions
+{
+  template <int dim, typename RangeNumberType>
+  class SymbolicFunction;
+}
+#endif
+
+namespace Functions
+{
+#ifdef DEAL_II_WITH_SYMENGINE
+  /**
+   * A Function class that leaverages symbolic differentiation to compute
+   * gradients, laplacians, hessians, and time derivatives.
+   *
+   * This class can be used to define functions using methods provided by the
+   * Differentiation::SD namespace. In particular, one can define a symbolic
+   * evaluation point (the argument of the function), as well as a symbolic
+   * expression.
+   *
+   * The symbolic gradients and the symbolic Hessians are computed at
+   * construction time, and when a substitution in the symbolic functions is
+   * requested by the user using the method apply_substitution_map().
+   *
+   * Whenever one of the evaluation methods is called, a substitution is
+   * attempted with the symbolic argument replaced by the evaluation point.
+   * The user has to make sure that at evaluation time argument substitution
+   * provides a fully evaluated expression (i.e., no other symbols are contained
+   * in the function expression, except numerical values), or an exception will
+   * be thrown. Additional symbols can be partially evaluated or substituted
+   * by calling apply_substitution_map().
+   *
+   * The simplest use case of this class is given in the following example:
+   * @code
+   * SymbolicFunction<2> fun("x^2+y; t*x*y");
+   * fun.set_time(3.0);
+   * Point<2> p(1.0, 2.0);
+   *
+   * auto a = fun.value(p, 0); // a = 3.0
```

Comment:
Maybe as a reminder for slow people like me:
```suggestion
   * auto a = fun.value(p, /*component=*/0); // a = 3.0
```

Reply:
How do you put a comment like `/* ... */` in the doxygen body of a code snippet? Mumble mumble...

Reply:
I think we've got places where we use `/ *` with a space in between. Visually, it's going to be nearly indistinguishable.